packages/SwingSet/src/vats/timer/vat-timer.js

```diff
@@ -1,98 +1,854 @@
 // @ts-check
 
-import { Nat } from '@agoric/nat';
-import { assert, details as X } from '@agoric/assert';
+import { E } from '@endo/eventual-send';
 import { Far, passStyleOf } from '@endo/marshal';
-import { makeNotifierFromAsyncIterable } from '@agoric/notifier';
 import { makePromiseKit } from '@endo/promise-kit';
-
-import { makeTimedIterable } from './timed-iteration.js';
+import { Nat } from '@agoric/nat';
+import { assert } from '@agoric/assert';
+import {
+  provideKindHandle,
+  defineDurableKind,
+  defineDurableKindMulti,
+  makeScalarBigMapStore,
+  makeScalarBigWeakMapStore,
+} from '@agoric/vat-data';
+import { provide, makeLegacyWeakMap } from '@agoric/store';
 import { TimeMath } from './timeMath.js';
 
-export function buildRootObject(vatPowers) {
+// RAM usage: O(number of outstanding delay() promises) +
+// O(number of notifiers??). setWakeup() only consumes DB.
+
+/**
+ * // Handler is a Far object with .wake(time) used for callbacks
+ *
+ * @typedef Handler
+ * @property {(scheduled: Timestamp) => unknown} wake
+ * @typedef {unknown} CancelToken
+ * // CancelToken must be pass-by-reference and durable, either local or remote
+ * @typedef Event
+ * @property {() => void} scheduleYourself
+ * @property {() => void} fired
+ * @property {() => void} cancel
+ * @typedef {MapStore<TimestampValue, Event[]>} Schedule
+ * @typedef {WeakMapStore<CancelToken, Event[]>} CancelTable
+ * @typedef {unknown} PromiseEvent
+ * @typedef WakeupPromiseControls
+ * @property {(scheduled: Timestamp) => void} resolve
+ * @property {(err: unknown) => void} reject
+ * @typedef {LegacyWeakMap<PromiseEvent, WakeupPromiseControls>} WakeupPromiseTable
+ */
+
+/* Repeaters have an Event with both 'start' and 'interval'. One-shot
+ * wakeups are Events with .interval = undefined. Each Event is
+ * tri-state: "scheduled" (.scheduled is non-undefined), "cancelled"
+ * (.cancelled=true), or "executing" (only for Repeaters, indicated by
+ * .scheduled=undefined and .cancelled = false). Events are kept alive
+ * by the schedule while "scheduled", by handler.wake result promise
+ * callbacks while "executing", and by mostly nothing when "cancelled".
+ */
+
+// these internal functions are exported for unit tests
```

Comment:
I'll add the "this is why this is safe" comment. They're pure functions. Also, they're exported under a separate name, so when this file is used as a vat entry point, we can leverage the fact that liveslots and the `subprocess-xsnap` supervisor don't touch anything other than `buildRootObject`.

For the benefit of the `manualTimer` unit-test support tool (used by zoe's tests, and others), I might need to add a non-pure portion which lets the unit test driver get a callback with every `wake()` result promise, so it can `await` them as it sees fit. I might also be able to get away without that.. still trying to look for solutions. If I do add that, we'll be relying upon the "only `buildRootObject`" aspect for security.